File: sdks/dart/core/test/client_test.dart
import 'dart:async';

import 'package:orvano_core/orvano_core.dart';
import 'package:test/test.dart';

import 'fake_orvano.dart';

void main() {
  FakeOrvano? running;
  FakeOrvano server() => running!;
  final warnings = <String>[];

  Future<void> serve(List<Answer> answers) async =>
      running = await FakeOrvano.start(answers);

  Orvano app({
    String? endpoint,
    String? project,
    SessionStore? session,
    int maxRetries = 3,
    Duration? timeout,
  }) => Orvano(
    Client(
      endpoint: endpoint ?? server().endpoint,
      project: project,
      session: session,
      maxRetries: maxRetries,
      timeout: timeout ?? const Duration(seconds: 5),
      onWarning: warnings.add,
    ),
  );

  setUp(warnings.clear);
  tearDown(() async {
    await running?.close();
    running = null;
  });

  group('headers', () {
    test('sends the SDK name and version on every request (AC-11)', () async {
      await serve([health()]);

      await app().health.get();

      expect(
        server().requests.single.value('X-Orvano-SDK'),
        'orvano_core/$sdkVersion',
      );
    });

    test(
      'sends the project and the session token, and never an API key (AC-4)',
      () async {
        await serve([health()]);

        await app(project: 'p1', session: MemorySessionStore('t')).health.get();

        final headers = server().requests.single;
        expect(headers.value('X-Orvano-Project'), 'p1');
        expect(headers.value('X-Orvano-Session'), 't');
        expect(headers.value('X-Orvano-Key'), isNull);
      },
    );

    test('joins the endpoint and the operation path with one slash', () async {
      await serve([health()]);

      await app(endpoint: '${server().endpoint}/').health.get();

      expect(server().paths.single, '/v1/health');
    });

    test('refuses an endpoint that is not an absolute URL', () {
      expect(() => Client(endpoint: '/orvano'), throwsArgumentError);
    });
  });

  group('version warning (AC-11)', () {
    test('warns once per client when the server runs another minor', () async {
      await serve([health('0.99.0')]);
      final orvano = app();

      await orvano.health.get();
      await orvano.health.get();

      expect(warnings, hasLength(1));
      expect(
        warnings.single,
        allOf(contains('runs 0.99.0'), contains('orvano_core 0.99.x')),
      );
    });

    test('stays quiet when only the patch version differs', () async {
      final [major, minor, ...] = sdkVersion.split('.');
      await serve([health('$major.$minor.99')]);

      await app().health.get();

      expect(warnings, isEmpty);
    });
  });

  group('retries (AC-14)', () {
    for (final code in [503, 429]) {
      test('retries a GET after a $code, honoring Retry-After', () async {
        await serve([status(code, retryAfter: '0'), health()]);

        final result = await app().health.get();

        expect(result.status, 'ok');
        expect(server().requests, hasLength(2));
      });
    }

    test('gives up after maxRetries with the last status', () async {
      await serve([status(503, retryAfter: '0')]);

      await expectLater(
        app(maxRetries: 2).health.get(),
        throwsA(isA<OrvanoException>().having((e) => e.status, 'status', 503)),
      );
      expect(server().requests, hasLength(3));
    });

    test('does not retry other failures', () async {
      await serve([status(500)]);

      await expectLater(app().health.get(), throwsA(isA<OrvanoException>()));
      expect(server().requests, hasLength(1));
    });

    test('never retries a POST that is not marked idempotent', () async {
      await serve([status(503, retryAfter: '0')]);

      await expectLater(
        app().client.send('POST', '/v1/things'),
        throwsA(isA<OrvanoException>()),
      );
      expect(server().methods, ['POST']);
    });

    test('retries a POST marked idempotent', () async {
      await serve([status(503, retryAfter: '0'), status(204)]);

      await app().client.send('POST', '/v1/things', idempotent: true);

      expect(server().methods, ['POST', 'POST']);
    });
  });

  group('timeouts (AC-14)', () {
    test('times out a call that takes too long', () async {
      await serve([hang()]);

      await expectLater(
        app().health.get(
          options: const RequestOptions(timeout: Duration(milliseconds: 100)),
        ),
        throwsA(isA<TimeoutException>()),
      );
    });

    test('covers retry waits with the same timeout', () async {
      await serve([status(503, retryAfter: '30')]);
      final watch = Stopwatch()..start();

      await expectLater(
        app(timeout: const Duration(milliseconds: 200)).health.get(),
        throwsA(isA<TimeoutException>()),
      );
      expect(watch.elapsed, lessThan(const Duration(seconds: 5)));
    });
  });

  group('errors (AC-6)', () {
    test(
      'maps a problem body to one exception with its code, detail, and request ID',
      () async {
        await serve([
          problem(409, {
            'title': 'Conflict',
            'status': 409,
            'detail': 'Already there.',
            'code': 'user_already_exists',
            'requestId': 'body-id',
          }, requestId: 'header-id'),
        ]);

        await expectLater(
          app().health.get(),
          throwsA(
            isA<OrvanoException>()
                .having((e) => e.status, 'status', 409)
                .having((e) => e.code, 'code', 'user_already_exists')
                .having((e) => e.message, 'message', 'Already there.')
                .having((e) => e.requestId, 'requestId', 'body-id'),
          ),
        );
      },
    );

    test('falls back to the title when a problem has no detail', () async {
      await serve([
        problem(404, {
          'title': 'Not Found',
          'status': 404,
          'code': 'not_found',
          'requestId': 'r',
        }),
      ]);

      await expectLater(
        app().health.get(),
        throwsA(
          isA<OrvanoException>().having(
            (e) => e.message,
            'message',
            'Not Found',
          ),
        ),
      );
    });

    test(
      'uses unknown and the request ID header when the body is not a problem',
      () async {
        await serve([
          (response) async {
            response
              ..statusCode = 502
              ..headers.set('X-Request-Id', 'header-id')
              ..write('<html>Bad gateway</html>');
            await response.close();
          },
        ]);

        await expectLater(
          app().health.get(),
          throwsA(
            isA<OrvanoException>()
                .having((e) => e.status, 'status', 502)
                .having((e) => e.code, 'code', 'unknown')
                .having((e) => e.requestId, 'requestId', 'header-id'),
          ),
        );
      },
    );

    test('exposes the generated error codes as constants', () {
      expect(
        (ErrorCode.notFound, ErrorCode.internalError),
        ('not_found', 'internal_error'),
      );
    });
  });

  group('paginate and events', () {
    test('paginate walks every page in order (AC-7)', () async {
      final pages = {
        null: (items: [1, 2], nextCursor: 'b'),
        'b': (items: [3, 4], nextCursor: 'c'),
        'c': (items: [5], nextCursor: null),
      };
      final cursors = <String?>[];

      final items =
          await paginate<({List<int> items, String? nextCursor}), int>((
            cursor,
          ) async {
            cursors.add(cursor);
            return pages[cursor]!;
          }, (page) => (page.items, page.nextCursor)).toList();

      expect(items, [1, 2, 3, 4, 5]);
      expect(cursors, [null, 'b', 'c']);
    });

    test(
      'decodeEvent returns null for an event this SDK does not know (AC-8)',
      () {
        expect(decodeEvent('nobody.knows', {'a': 1}), isNull);
      },
    );

    test('decodeEvent decodes through the registry it is given (AC-8)', () {
      final decoded = decodeEvent(
        'things.created',
        {'id': 't1'},
        registry: {'things.created': (json) => json['id'] as String},
      );

      expect(decoded, 't1');
    });
  });
}

File: sdks/js/test/client.test.ts
import { describe, expect, it, vi } from 'vitest'

import { Client, MemorySessionStore, Orvano, OrvanoError, sdkVersion } from '../src/index.js'
import type { ClientConfig } from '../src/index.js'
import { Client as ServerClient, Orvano as ServerOrvano } from '../src/server.js'
import { fakeFetch, health, problem, status } from './fake-fetch.js'

const endpoint = 'https://orvano.example.com'
const quiet = { warn: vi.fn() }

function app(fetch: typeof globalThis.fetch, config: Partial<ClientConfig> = {}): Orvano {
  return new Orvano(new Client({ endpoint, fetch, logger: quiet, ...config }))
}

describe('headers', () => {
  it('sends the SDK name and version on every request (AC-11)', async () => {
    const { fetch, sent } = fakeFetch(health())

    await app(fetch).health.get()

    expect(sent[0]?.headers.get('X-Orvano-SDK')).toBe(`@orvano/js/${sdkVersion}`)
  })

  it('sends the project and the session token from the session store', async () => {
    const { fetch, sent } = fakeFetch(health())

    await app(fetch, { project: 'p1', session: new MemorySessionStore('t') }).health.get()

    expect(sent[0]?.headers.get('X-Orvano-Project')).toBe('p1')
    expect(sent[0]?.headers.get('X-Orvano-Session')).toBe('t')
  })

  it('never sends an API key from the app entry (AC-4)', async () => {
    const { fetch, sent } = fakeFetch(health())

    await app(fetch, { headers: {} }).health.get()

    expect(sent[0]?.headers.has('X-Orvano-Key')).toBe(false)
    expect(sent[0]?.headers.has('X-Orvano-Session')).toBe(false)
  })

  it('sends the API key from the server entry (AC-4)', async () => {
    const { fetch, sent } = fakeFetch(health())

    await new ServerOrvano(
      new ServerClient({ endpoint, fetch, logger: quiet, apiKey: 'test-server-key' }),
    ).health.get()

    expect(sent[0]?.headers.get('X-Orvano-Key')).toBe('test-server-key')
  })

  it.each([endpoint, `${endpoint}/`])(
    'joins %s and the operation path with one slash',
    async (base) => {
      const { fetch, sent } = fakeFetch(health())

      await app(fetch, { endpoint: base }).health.get()

      expect(sent[0]?.url).toBe(`${endpoint}/v1/health`)
    },
  )

  it('refuses an endpoint that is not an absolute URL', () => {
    expect(() => new Client({ endpoint: '/orvano' })).toThrow(TypeError)
  })

  it('leaves undefined query values out of the URL', async () => {
    const { fetch, sent } = fakeFetch(health())

    await new Client({ endpoint, fetch, logger: quiet }).request({
      method: 'GET',
      path: '/v1/things',
      query: { limit: 2, cursor: undefined },
    })

    expect(sent[0]?.url).toBe(`${endpoint}/v1/things?limit=2`)
  })
})

describe('version warning (AC-11)', () => {
  it('warns once per client when the server runs another minor', async () => {
    const { fetch } = fakeFetch(health('0.99.0'))
    const warn = vi.fn()
    const orvano = app(fetch, { logger: { warn } })

    await orvano.health.get()
    await orvano.health.get()

    expect(warn).toHaveBeenCalledTimes(1)
    expect(warn.mock.calls[0]?.[0]).toContain('runs 0.99.0')
    expect(warn.mock.calls[0]?.[0]).toContain('@orvano/js 0.99.x')
  })

  it('stays quiet when only the patch version differs', async () => {
    const [major, minor] = sdkVersion.split('.')
    const { fetch } = fakeFetch(health(`${major ?? '0'}.${minor ?? '0'}.99`))
    const warn = vi.fn()

    await app(fetch, { logger: { warn } }).health.get()

    expect(warn).not.toHaveBeenCalled()
  })

  it('stays quiet when the server sends no version', async () => {
    const { fetch } = fakeFetch(() => Response.json({ status: 'ok', version: 'x' }))
    const warn = vi.fn()

    await app(fetch, { logger: { warn } }).health.get()

    expect(warn).not.toHaveBeenCalled()
  })
})

describe('retries (AC-14)', () => {
  it.each([503, 429])('retries a GET after a %i, honoring Retry-After', async (code) => {
    const { fetch, sent } = fakeFetch(status(code, '0'), health())

    const result = await app(fetch).health.get()

    expect(result.status).toBe('ok')
    expect(sent).toHaveLength(2)
  })

  it('gives up after maxRetries with the last status', async () => {
    const { fetch, sent } = fakeFetch(status(503, '0'))

    await expect(app(fetch, { maxRetries: 2 }).health.get()).rejects.toMatchObject({ status: 503 })
    expect(sent).toHaveLength(3)
  })

  it('does not retry other failures', async () => {
    const { fetch, sent } = fakeFetch(status(500))

    await expect(app(fetch).health.get()).rejects.toBeInstanceOf(OrvanoError)
    expect(sent).toHaveLength(1)
  })

  it('never retries a POST that is not marked idempotent', async () => {
    const { fetch, sent } = fakeFetch(status(503, '0'))

    await expect(
      new Client({ endpoint, fetch, logger: quiet }).request({
        method: 'POST',
        path: '/v1/things',
      }),
    ).rejects.toMatchObject({ status: 503 })
    expect(sent).toHaveLength(1)
  })

  it('retries a POST marked idempotent', async () => {
    const { fetch, sent } = fakeFetch(status(503, '0'), status(204))

    await new Client({ endpoint, fetch, logger: quiet }).request({
      method: 'POST',
      path: '/v1/things',
      idempotent: true,
    })

    expect(sent).toHaveLength(2)
  })
})

describe('timeouts and cancellation (AC-14)', () => {
  it('times out a call that takes too long', async () => {
    const { fetch } = fakeFetch('hang')

    await expect(app(fetch).health.get({ timeoutMs: 50 })).rejects.toMatchObject({
      name: 'TimeoutError',
    })
  })

  it('stops at once when the caller cancels', async () => {
    const { fetch } = fakeFetch('hang')
    const cancel = new AbortController()
    cancel.abort()

    await expect(app(fetch).health.get({ signal: cancel.signal })).rejects.toMatchObject({
      name: 'AbortError',
    })
  })
})

describe('errors (AC-6)', () => {
  it('maps a problem body to one error with its code, detail, and request ID', async () => {
    const { fetch } = fakeFetch(
      problem(
        409,
        {
          title: 'Conflict',
          status: 409,
          detail: 'Already there.',
          code: 'user_already_exists',
          requestId: 'body-id',
        },
        'header-id',
      ),
    )

    await expect(app(fetch).health.get()).rejects.toMatchObject({
      name: 'OrvanoError',
      status: 409,
      code: 'user_already_exists',
      message: 'Already there.',
      requestId: 'body-id',
    })
  })

  it('falls back to the title when a problem has no detail', async () => {
    const { fetch } = fakeFetch(
      problem(404, { title: 'Not Found', status: 404, code: 'not_found', requestId: 'r' }),
    )

    await expect(app(fetch).health.get()).rejects.toMatchObject({
      code: 'not_found',
      message: 'Not Found',
    })
  })

  it('uses unknown and the request ID header when the body is not a problem', async () => {
    const { fetch } = fakeFetch(
      () =>
        new Response('<html>Bad gateway</html>', {
          status: 502,
          headers: { 'X-Request-Id': 'header-id' },
        }),
    )

    await expect(app(fetch).health.get()).rejects.toMatchObject({
      status: 502,
      code: 'unknown',
      requestId: 'header-id',
    })
  })
})
